File: core/src/main/java/com/views/progressloadingview/window/WP10ProgressBar.java
package com.views.progressloadingview.window;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.Gravity;
import android.widget.RelativeLayout;

import androidx.annotation.Nullable;

import com.R;
import com.core.utilities.LAppResource;

import java.util.ArrayList;
//TODO convert kotlin
public class WP10ProgressBar extends RelativeLayout {

    private static final int INTERVAL_DEF = 150;
    private static final int INDICATOR_COUNT_DEF = 5;
    private static final int ANIMATION_DURATION_DEF = 1800;
    private static final int INDICATOR_HEIGHT_DEF = 7;
    private static final int INDICATOR_RADIUS_DEF = 0;

    private int interval;
    private int animationDuration;
    private int indicatorHeight;
    private int indicatorColor;
    private int indicatorRadius;

    private boolean isShowing = false;
    private ArrayList<WP10Indicator> wp10Indicators;

    private Handler handler;
    private int progressBarCount = 0;

    public WP10ProgressBar(Context context) {
        super(context);
        initialize(null);
    }

    public WP10ProgressBar(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initialize(attrs);
    }

    public WP10ProgressBar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize(attrs);
    }

    private void initialize(@Nullable AttributeSet attrs) {
        this.setGravity(Gravity.CENTER);
        this.handler = new Handler();
        this.setRotation(-25);
        setAttributes(attrs);
        initializeIndicators();
    }

    private void setAttributes(AttributeSet attributeSet) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attributeSet, R.styleable.WP10ProgressBar);
        interval = typedArray.getInt(R.styleable.WP7ProgressBar_wpInterval, INTERVAL_DEF);
        animationDuration = typedArray.getInt(R.styleable.WP7ProgressBar_wpAnimationDuration, ANIMATION_DURATION_DEF);
        indicatorHeight = typedArray.getInt(R.styleable.WP7ProgressBar_wpIndicatorHeight, INDICATOR_HEIGHT_DEF);
        indicatorRadius = typedArray.getInt(R.styleable.WP7ProgressBar_wpIndicatorRadius, INDICATOR_RADIUS_DEF);
        indicatorColor = typedArray.getColor(R.styleable.WP7ProgressBar_wpIndicatorColor, LAppResource.INSTANCE.getColor(R.color.colorAccent));
        typedArray.recycle();
    }

    private void showAnimation() {
        for (int i = 0; i < wp10Indicators.size(); i++) {
            wp10Indicators.get(i).startAnim(animationDuration, (5 - i) * interval);
        }
    }

    private void initializeIndicators() {
        this.removeAllViews();
        ArrayList<WP10Indicator> WP10Indicators = new ArrayList<>();
        for (int i = 0; i < INDICATOR_COUNT_DEF; i++) {
            WP10Indicator wp10Indicator = new WP10Indicator(getContext(), indicatorHeight, indicatorColor, indicatorRadius, i);
//                wp10Indicator.setRotation(i * 14);
            WP10Indicators.add(wp10Indicator);
            this.addView(wp10Indicator);
        }
        this.wp10Indicators = WP10Indicators;
    }

    private void show() {
        if (isShowing)
            return;
        isShowing = true;
        showAnimation();
//        startWholeViewAnimation();
    }

    private void hide() {
        clearIndicatorsAnimations();
        isShowing = false;
//        hideWholeViewAnimation();
    }

    private void clearIndicatorsAnimations() {
        for (WP10Indicator wp10Indicator : wp10Indicators) {
            wp10Indicator.removeAnim();
        }
    }

    public void showProgressBar() {
        progressBarCount++;
        if (progressBarCount == 1) {
            handler.post(new Runnable() {
                @Override
                public void run() {
                    WP10ProgressBar.this.show();
                }
            });
        }
    }

    public void hideProgressBar() {
        progressBarCount--;
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (progressBarCount == 0) {
                    WP10ProgressBar.this.hide();
                }
            }
        }, 50);
    }

    public void setInterval(int interval) {
        this.interval = interval;
        initializeIndicators();
    }

    public void setAnimationDuration(int animationDuration) {
        this.animationDuration = animationDuration;
        initializeIndicators();
    }

    public void setIndicatorHeight(int indicatorHeight) {
        this.indicatorHeight = indicatorHeight;
        initializeIndicators();
    }

    public void setIndicatorColor(int indicatorColor) {
        this.indicatorColor = indicatorColor;
        initializeIndicators();
    }

    public void setIndicatorRadius(int indicatorRadius) {
        this.indicatorRadius = indicatorRadius;
        initializeIndicators();
    }
}

File: core/src/main/java/com/loitp/views/loading/window/WP10ProgressBar.kt
package com.loitp.views.loading.window

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.Gravity
import android.widget.RelativeLayout
import com.loitp.R
import com.loitp.core.ext.getColor

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class WP10ProgressBar : RelativeLayout {

    companion object {
        private const val INTERVAL_DEF = 150
        private const val INDICATOR_COUNT_DEF = 5
        private const val ANIMATION_DURATION_DEF = 1800
        private const val INDICATOR_HEIGHT_DEF = 7
        private const val INDICATOR_RADIUS_DEF = 0
    }

    private var interval = 0
    private var animationDuration = 0
    private var indicatorHeight = 0
    private var indicatorColor = 0
    private var indicatorRadius = 0
    private var isShowing = false
    private var listWP10Indicators: ArrayList<WP10Indicator>? = null
    private var mHandler: Handler? = null
    private var progressBarCount = 0

    constructor(context: Context) : super(context) {
        initialize(null)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?
    ) : super(context, attrs) {
        initialize(attrs)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int
    ) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        initialize(attrs)
    }

    private fun initialize(attrs: AttributeSet?) {
        this.gravity = Gravity.CENTER
        mHandler = Handler(Looper.getMainLooper())
        this.rotation = -25f
        setAttributes(attrs)
        initializeIndicators()
    }

    private fun setAttributes(attributeSet: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(attributeSet, R.styleable.WP10ProgressBar)
        interval = typedArray.getInt(R.styleable.WP7ProgressBar_wpInterval, INTERVAL_DEF)
        animationDuration = typedArray.getInt(
            R.styleable.WP7ProgressBar_wpAnimationDuration,
            ANIMATION_DURATION_DEF
        )
        indicatorHeight =
            typedArray.getInt(R.styleable.WP7ProgressBar_wpIndicatorHeight, INDICATOR_HEIGHT_DEF)
        indicatorRadius =
            typedArray.getInt(R.styleable.WP7ProgressBar_wpIndicatorRadius, INDICATOR_RADIUS_DEF)
        indicatorColor = typedArray.getColor(
            R.styleable.WP7ProgressBar_wpIndicatorColor,
            getColor(R.color.colorAccent)
        )
        typedArray.recycle()
    }

    private fun showAnimation() {
        listWP10Indicators?.let {
            for (i in it.indices) {
                it[i].startAnim(
                    animationDuration = animationDuration.toLong(),
                    delay = ((5 - i) * interval).toLong()
                )
            }
        }
    }

    private fun initializeIndicators() {
        removeAllViews()
        val wP10Indicators = ArrayList<WP10Indicator>()
        for (i in 0 until INDICATOR_COUNT_DEF) {
            val wp10Indicator = WP10Indicator(
                context = context,
                indicatorHeight = indicatorHeight,
                color = indicatorColor,
                radius = indicatorRadius,
                number = i
            )
            wP10Indicators.add(wp10Indicator)
            this.addView(wp10Indicator)
        }
        listWP10Indicators = wP10Indicators
    }

    private fun show() {
        if (isShowing) return
        isShowing = true
        showAnimation()
    }

    private fun hide() {
        clearIndicatorsAnimations()
        isShowing = false
    }

    private fun clearIndicatorsAnimations() {
        listWP10Indicators?.forEach {
            it.removeAnim()
        }
    }

    fun showProgressBar() {
        progressBarCount++
        if (progressBarCount == 1) {
            mHandler?.post {
                show()
            }
        }
    }

    fun hideProgressBar() {
        progressBarCount--
        mHandler?.postDelayed({
            if (progressBarCount == 0) {
                hide()
            }
        }, 50)
    }

    @Suppress("unused")
    fun setInterval(interval: Int) {
        this.interval = interval
        initializeIndicators()
    }

    @Suppress("unused")
    fun setAnimationDuration(animationDuration: Int) {
        this.animationDuration = animationDuration
        initializeIndicators()
    }

    @Suppress("unused")
    fun setIndicatorHeight(indicatorHeight: Int) {
        this.indicatorHeight = indicatorHeight
        initializeIndicators()
    }

    @Suppress("unused")
    fun setIndicatorColor(indicatorColor: Int) {
        this.indicatorColor = indicatorColor
        initializeIndicators()
    }

    fun setIndicatorRadius(indicatorRadius: Int) {
        this.indicatorRadius = indicatorRadius
        initializeIndicators()
    }
}
